src/filesystem.cpp

```diff
@@ -685,194 +670,83 @@ static std::optional<std::string> get_games_path()
+	// if no custom userdata directory was provided, use appropriate default
+	if(newprefdir.empty()) {
+#ifdef _WIN32
+		newprefdir = "~/Wesnoth" + get_version_path_suffix();
+#elif defined(__APPLE__)
+		newprefdir = "~/Library/Application Support/Wesnoth_"+get_version_path_suffix();
+#elif defined(WESNOTH_BOOST_OS_IOS)
+		char* sdl_pref_path = SDL_GetPrefPath("wesnoth.org", "iWesnoth");
+		if(sdl_pref_path) {
+			newprefdir = std::string(sdl_pref_path);
+			SDL_free(sdl_pref_path);
 		} else {
-			dir = get_cwd();
+			newprefdir = "~/.wesnoth" + get_version_path_suffix();
 		}
-		dir /= temp.substr(1);
-	} else {
-		if(temp.empty()) {
-			DBG_FS << "using default userdata folder name";
-			temp = "Wesnoth" + get_version_path_suffix();
+#else
+		const char* h = std::getenv("HOME");
+		std::string home = h ? h : "";
+		h = std::getenv("XDG_DATA_HOME");
+		std::string xdg_data_home = h ? h : "";
+		if (!xdg_data_home.empty()) {
+			newprefdir = xdg_data_home + "/wesnoth/" + get_version_path_suffix();
+		} else if (!home.empty()) {
+			newprefdir = home + "/.local/share/wesnoth/" + get_version_path_suffix();
 		} else {
-			ERR_FS << "relative path for userdata whose first path component is neither '.' or '..' is not allowed: " << temp
-				   << ", using default location under My Games";
-			temp = "Wesnoth" + get_version_path_suffix();
+			newprefdir = ".wesnoth" + get_version_path_suffix();
 		}
+#endif
+	}
 
+	bfs::path dir;
+	if(newprefdir[0] == '~') {
+#ifdef _WIN32
 		std::optional<std::string> games_path = get_games_path();
 		if(games_path) {
 			create_directory_if_missing(*games_path);
 			dir = *games_path;
-			dir /= temp;
-			DBG_FS << "userdata is under My Games at " << dir.string();
 		} else {
-			//
-			// Crummy fallback path full of pain and suffering.
-			//
 			dir = get_cwd();
-			dir /= temp;
-			DBG_FS << "userdata is at " << dir.string();
-		}
-	}
-
-	return dir;
-}
-#elif defined(__APPLE__) || defined(WESNOTH_BOOST_OS_IOS)
-static bfs::path apple_userdata(const std::string& newprefdir)
-{
-	bfs::path dir;
-	std::string temp = newprefdir;
-	const char* home_str = getenv("HOME");
-
-	// if a custom userdata was not specified
-	// use the PREFERENCES_DIR if defined
-	// else if this is iOS, use the SDL pref path
-	// else use a default - this is currently the "unsandboxed" location from migrate_apple_config_directory_for_unsandboxed_builds() above
-	if(temp.empty()) {
-#ifdef PREFERENCES_DIR
-		temp = PREFERENCES_DIR;
-		DBG_FS << "userdata using PREFERENCES_DIR '" << PREFERENCES_DIR << "'";
-		if(home_str && temp[0] == '~') {
-			temp = home_str + temp.substr(1);
-			DBG_FS << "userdata is relative to HOME";
-		}
-#elif defined(WESNOTH_BOOST_OS_IOS)
-		char *sdl_pref_path = SDL_GetPrefPath("wesnoth.org", "iWesnoth");
-		if(sdl_pref_path) {
-			temp = std::string(sdl_pref_path) + ".wesnoth" + get_version_path_suffix();
-			SDL_free(sdl_pref_path);
+			WRN_FS << "Using current directory instead: " << dir.string();
```

Comment:
Instead of what? This codepath seems to be lacking any previous context from a WRN_FS or ERR_FS log.